tests/test-update-from-lan.c

```diff
@@ -0,0 +1,187 @@
+  g_autoptr(GFile) autoupdater_root = NULL;
+  g_autoptr(EosTestAutoupdater) autoupdater = NULL;
+  g_auto(CmdStuff) reaped = CMD_STUFF_CLEARED;
+  g_autoptr(GPtrArray) cmds = NULL;
+  g_autoptr(GPtrArray) cmds_to_free = NULL;
+  gboolean has_commit;
+  g_autoptr(GDateTime) client_timestamp = NULL;
+
+  server_root = g_file_get_child (fixture->tmpdir, "main");
+  server = eos_test_server_new_quick (server_root,
+                                      lan_server_count + 1,
+                                      default_vendor,
+                                      default_product,
+                                      FALSE,
+                                      default_ref,
+                                      0,
+                                      keyid,
+                                      default_ostree_path,
+                                      &error);
+  g_assert_no_error (error);
+  g_assert_cmpuint (server->subservers->len, ==, 1u);
+
+  subserver = g_object_ref (EOS_TEST_SUBSERVER (g_ptr_array_index (server->subservers, 0)));
+  client_root = g_file_get_child (fixture->tmpdir, "client");
+  client = eos_test_client_new (client_root,
+                                default_remote_name,
+                                subserver,
+                                default_ref,
+                                default_vendor,
+                                default_product,
+                                &error);
+  g_assert_no_error (error);
+
+  lan_servers = object_array_new ();
+  lan_server_cmds = g_ptr_array_new_with_free_func ((GDestroyNotify)cmd_async_stuff_free);
+  for (idx = 0; idx < lan_server_count; ++idx)
+    {
+      g_autofree gchar *dir_name = g_strdup_printf ("lan_server_%u", idx);
+      g_autoptr(GFile) lan_server_root = g_file_get_child (fixture->tmpdir, dir_name);
+      g_autoptr(EosTestClient) lan_server = NULL;
+      g_autoptr(GKeyFile) definition = NULL;
+      g_autoptr(CmdAsyncStuff) server_cmd = NULL;
+
+      g_date_time_unref (subserver->branch_file_timestamp);
+      subserver->branch_file_timestamp = days_ago (lan_server_count - idx);
+      g_hash_table_insert (subserver->ref_to_commit,
+                           g_strdup (default_ref),
+                           GUINT_TO_POINTER (1 + idx));
+      eos_test_subserver_update (subserver,
+                                 &error);
+      g_assert_no_error (error);
+      lan_server = eos_test_client_new (lan_server_root,
+                                        default_remote_name,
+                                        subserver,
+                                        default_ref,
+                                        default_vendor,
+                                        default_product,
+                                        &error);
+      g_assert_no_error (error);
+      server_cmd = g_new0 (CmdAsyncStuff, 1);
+      eos_test_client_run_update_server (lan_server,
+                                         12345 + idx,
```

Comment:
I modified it later to actually pick up a random free port instead of hardcoding it.

https://github.com/kinvolk/eos-updater/commit/f9d423f17cd7362a62438c35baa906006d35c924 and https://github.com/kinvolk/eos-updater/commit/628f8ba22098064e9129f8990df218e27a3b5e2c